src/infill/ZigzagConnectorProcessorConnectedEndPieces.cpp

```diff
@@ -20,50 +20,54 @@ void ZigzagConnectorProcessorConnectedEndPieces::registerScanlineSegmentIntersec
     }
     else
     {
-        if (previous_scanline_is_even)
-        { // when a boundary segment starts in an even scanline it is either a normal zigzag connector or an endpiece, so it should be included anyway
-            addLine(last_connector_point, intersection);
-        }
-        else if (!previous_scanline_is_even && !this_scanline_is_even) // if we end an odd boundary in an odd segment
-        { // add whole zigzag_connector (including the just obtained point)
-            for (unsigned int point_idx = 1; point_idx < zigzag_connector.size(); point_idx++)
+        bool add_connection = previous_scanline_is_even;  // add connections on even segments
+        add_connection |= previous_scanline_is_even == this_scanline_is_even;  // end piece
+
+        if (add_connection)
+        {
+            if (skip_some_zags && ++current_zag_count >= zag_skip_count)
             {
-                addLine(zigzag_connector[point_idx - 1], zigzag_connector[point_idx]);
+                // skip this zag and reset the count
+                current_zag_count = 0;
+            }
+            else
+            {
+                for (unsigned int point_idx = 1; point_idx < zigzag_connector.size(); point_idx++)
+                {
+                    addLine(zigzag_connector[point_idx - 1], zigzag_connector[point_idx]);
+                }
+                addLine(zigzag_connector.back(), intersection);
             }
-            addLine(zigzag_connector.back(), intersection);
-            zigzag_connector.clear();
         }
-
     }
     zigzag_connector.clear(); // we're starting a new (odd) zigzag connector, so clear the old one
-    if (!this_scanline_is_even) // we are either in an end piece or an boundary segment starting in an odd scanline
-    { // only when a boundary segment starts in an odd scanline it depends on whether it ends in an odd scanline for whether this segment should be included or not
-        zigzag_connector.push_back(intersection);
-    }
-
+    zigzag_connector.push_back(intersection);
     last_connector_point = intersection;
 }
 
 
 void ZigzagConnectorProcessorConnectedEndPieces::registerPolyFinished()
 {
-    // write end segment if needed (first half of start/end-crossing segment)
-    if (!last_scanline_is_even && !first_zigzag_connector_ends_in_even_scanline)
+    const bool is_last_piece_end_piece = last_scanline_is_even == first_zigzag_connector_ends_in_even_scanline;
+    const bool need_to_skip_this_piece = skip_some_zags && ++current_zag_count >= zag_skip_count;
+    const bool add_last_piece = is_last_piece_end_piece || (last_scanline_is_even && !need_to_skip_this_piece);
+
+    if (add_last_piece)
```

Comment:
I always find it easier to read when it's like this:

    if (<condition>)
    {
        <do stuff>
    }

Rather than:

    const bool should_do_stuff = <condition>;
    if (should_do_stuff)
    {
        <do stuff>
    }

It's sort of what if-clauses are made for.